include/memory/memory.hh

```diff
@@ -33,20 +129,145 @@ public:
+    throw PhysMemory::MisAlignedAddrException(
+        "Misaligned memory access is not supported!");
+  AddrSections sections(addr);
+  auto offset = sections.offset;
+  auto isInTLB = tlb.tlbLookup(addr);
+  PagePtr page;
+  if (isInTLB.has_value()) {
+    page = isInTLB.value();
+  } else {
+    page = PhysMemory::pageTableLookup<op>(sections);
+    tlb.tlbUpdate(addr, page);
+  }
+  Word *word = &page->wordStorage.at(offset / sizeof(Word));
+  Byte *byte = reinterpret_cast<Byte *>(word) + (offset % sizeof(Word));
+  return reinterpret_cast<T *>(byte);
+}
+
+template <PhysMemory::MemoryOp op>
+PagePtr PhysMemory::pageTableLookup(const AddrSections &sect) {
+
+  using MemOp = PhysMemory::MemoryOp;
+  auto index = sect.indexPt;
+  auto it_PT = pageTable.find(index);
+  if (it_PT == pageTable.end()) {
+    if constexpr (op == MemOp::LOAD)
+      throw PhysMemory::PageFaultException(
+          "Load on unmapped region in physical mem");
+    else {
+      pageTable[index] = Page();
+    }
+  }
+  return &pageTable.at(index);
+}
+
+template <std::forward_iterator It>
+void Memory::storeRange(Addr start, It begin, It end) {
+  std::for_each(begin, end, [&start, this](auto curWord) {
+    storeWord(start, curWord);
+    start += kXLENInBytes;
+  });
+}
+
+inline void Memory::printMemStats(std::ostream &ost) const {
+  ost << "Memory stats:" << std::endl;
+  ost << "Loads: " << stats.numLoads << std::endl;
+  ost << "Stores: " << stats.numStores << std::endl;
+  ost << "End of memory stats." << std::endl;
+}
+
+inline const Memory::MemoryStats &Memory::getMemStats() const { return stats; }
+
+inline uint16_t PhysMemory::getOffset(Addr addr) {
+  return static_cast<uint16_t>(getBits<kOffsetBits - 1, 0>(addr));
+}
+
+inline Word Memory::loadWord(Addr addr) {
+  stats.numLoads++;
+  Word loadedWord = *physMem.getEntity<Word, PhysMemory::MemoryOp::LOAD>(addr);
+
+  return loadedWord;
+}
+
+inline void Memory::storeWord(Addr addr, Word word) {
+  stats.numStores++;
+  *physMem.getEntity<Word, PhysMemory::MemoryOp::STORE>(addr) = word;
+}
+
+inline Half Memory::loadHalf(Addr addr) {
+  stats.numLoads++;
+  Half loadedWord = *physMem.getEntity<Half, PhysMemory::MemoryOp::LOAD>(addr);
+
+  return loadedWord;
+}
+
+inline void Memory::storeHalf(Addr addr, Half half) {
+  stats.numStores++;
+  *physMem.getEntity<Half, PhysMemory::MemoryOp::STORE>(addr) = half;
+}
+
+inline Byte Memory::loadByte(Addr addr) {
+  stats.numLoads++;
+  Byte loadedWord = *physMem.getEntity<Byte, PhysMemory::MemoryOp::LOAD>(addr);
+
+  return loadedWord;
+}
+
+inline void Memory::storeByte(Addr addr, Byte byte) {
+  stats.numStores++;
+  *physMem.getEntity<Byte, PhysMemory::MemoryOp::STORE>(addr) = byte;
+}
+
+inline TLB::TLBIndex TLB::getTLBIndex(Addr addr) {
+  return static_cast<TLB::TLBIndex>(
+      getBits<(kTLBBits + kOffsetBits - 1), kOffsetBits>(addr));
+}
+
+inline std::optional<PagePtr> TLB::tlbLookup(Addr addr) {
```

Comment:
`std::optional` for pointer type looks like an overkill, because pointer already has special poison value `nullptr`